sw/device/mask_rom/mask_rom_start.S

```diff
@@ -230,6 +225,9 @@ _mask_rom_start_boot:
   la   a1, _bss_end
   call crt_section_clear
 
+  // Setup global pointer.
+  call crt_global_pointer_init
+
```

Comment:
Just out of curiosity, can this call be moved even further up?

Reply:
For my own benefit, which operations require `norelax`? Obviously setting the Global Pointer itself, but what else?

Reply:
> Just out of curiosity, can this call be moved even further up?

Yes, it just has to be after any bulk register clearing (which would 0 out `gp`) and before we jump into C code. Other than that it can go anywhere.

> which operations require `norelax`?

Linker relaxation is not something I've dealt with before so take this answer with a pinch of salt. As far as I know there aren't many other places we need to disable it:

* **Addressing global data:** The linker might replace instructions that address symbols in the `.sdata` section with instructions using `gp` relative addressing (I think maybe only if the relocations are relative to `__global_pointer$` though). So we would need to disable linker relaxations to access/address this data before the `gp` is set. I don't think this affects the `_data_start` and `_data_end` symbols but we might want to initialize the `gp` before using them anyway since it doesn't hurt.

* **Jump tables:** The other thing the linker might do is optimize `auipc` + `jalr` instruction sequences into `jal` (or even `c.jal`) instructions where the offset is known to fit in 21 bits. I think this only really affects jump tables such as the interrupt vector where we want each entry to be at a known offset from the symbol, so we need to disable linker relaxation for that.